src/common/transformations/src/transformations/common_optimizations/matmul_split_decomposition.cpp

```diff
@@ -0,0 +1,209 @@
+//
+
+#include "transformations/common_optimizations/matmul_split_decomposition.hpp"
+
+#include "itt.hpp"
+#include "openvino/core/rt_info.hpp"
+#include "openvino/opsets/opset1.hpp"
+#include "openvino/pass/pattern/op/or.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+
+using namespace ov;
+using namespace ov::pass::pattern;
+
+bool pass::MatmulGatherDecomposition::split_weights(const Output<Node>& weights,
+                                                    OutputVector& new_weights,
+                                                    Output<Node>* bias,
+                                                    OutputVector& new_bias,
+                                                    const bool transpose_b) {
+    // weights is static
+    if (weights.get_partial_shape().size() != 2u) {
+        return false;
+    }
+
+    if (bias) {
+        const auto& bias_rank = bias->get_partial_shape().size();
+        if (bias_rank != 3 && bias_rank != 1) {
+            return false;
+        }
+    }
+
+    // Decompose weights
+    auto axis = register_new_node(op::v0::Constant::create(element::i32, Shape{}, {transpose_b ? 0 : 1}));
+    auto split = register_new_node<op::v1::Split>(weights, axis, decompose_num);
+    for (auto& out : split->outputs()) {
+        new_weights.emplace_back(out);
+    }
+
+    if (bias) {
+        // Decompose bias
+        auto axis2 = register_new_node(op::v0::Constant::create(element::i32, Shape{}, {-1}));  // axis -1
+        auto split2 = register_new_node<op::v1::Split>(*bias, axis2, decompose_num);
+        for (auto& out : split2->outputs()) {
+            new_bias.emplace_back(out);
+        }
+    }
+    return true;
+}
+
+pass::MatmulGatherDecomposition::MatmulGatherDecomposition() {
+    MATCHER_SCOPE(MatmulGatherDecomposition);
+    auto input_pattern = any_input();
+    auto matmul_pattern = wrap_type<opset1::MatMul>({input_pattern, any_input(pattern::has_static_shape())},
+                                                    ov::pass::pattern::consumers_count(1));
+
+    auto bias_pattern = wrap_type<opset1::Constant>();
+    auto add_pattern = wrap_type<opset1::Add>({matmul_pattern, bias_pattern}, ov::pass::pattern::consumers_count(1));
+
+    auto reshape_productor_pattern = std::make_shared<pattern::op::Or>(OutputVector{matmul_pattern, add_pattern});
+
+    // Heuristics: Rank == 5, Baichun also match this pattern, but it only has rank 4, and have performance regression,
+    // so filter it out.
+    auto reshape_pattern =
+        wrap_type<opset1::Reshape>({reshape_productor_pattern, any_input()}, ov::pass::pattern::rank_equals(5));
+
+    // Heuristics: there should be only decompose_num(3) gathers to split
+    auto transpose_pattern =
+        wrap_type<opset1::Transpose>({reshape_pattern, ov::pass::pattern::wrap_type<ov::opset1::Constant>()},
+                                     ov::pass::pattern::consumers_count(decompose_num));
+    auto reshape2_pattern =
```

Comment:
probably we can use `pattern::optional` (https://github.com/openvinotoolkit/openvino/blob/master/src/core/include/openvino/pass/pattern/op/optional.hpp) here to simplify the pattern logic
auto reshape_pattern =  wrap_type<opset1::Reshape>(...)
auto optional_transpose = pattern::optional<opset1::Transpose>(...)
auto reshape2_pattern = wrap_type<opset1::Reshape>({optional_transpose, ...}...)

example:
https://github.com/openvinotoolkit/openvino/blob/7cf05641b6b1b249904c96e02ac07ee384219bb4/src/core/tests/pattern.cpp#L686-L720

Reply:
Updated